services/queue/src/artifacts.js

```diff
@@ -358,6 +417,102 @@ builder.declare({
+  name: 'finishArtifact',
+  stability: APIBuilder.stability.experimental,
+  category: 'Artifacts',
+  scopes: 'queue:create-artifact:<taskId>/<runId>',
+  input: 'finish-artifact-request.json#',
+  title: 'Finish Artifact',
+  description: [
+    'This endpoint marks an artifact as present for the given task, and',
+    'should be called when the artifact data is fully uploaded.',
+    '',
+    'The storage types `reference`, `link`, and `error` do not need to',
+    'be finished, as they are finished immediately by `createArtifact`.',
+    'The storage type `s3` does not support this functionality and cannot',
+    'be finished.  In all such cases, calling this method is an input error',
+    '(400).',
+  ].join('\n'),
+}, async function(req, res) {
+  const taskId = req.params.taskId;
+  const runId = parseInt(req.params.runId, 10);
+  const name = req.params.name;
+  const { uploadId } = req.body;
+
+  const [[artifact_row], task] = await Promise.all([
+    this.db.fns.get_queue_artifact(taskId, runId, name),
+    Task.get(this.db, taskId),
+  ]);
+
+  if (!artifact_row || !task || !task.runs[runId]) {
+    return res.reportError(
+      'ResourceNotFound',
+      'Artifact with taskId `{{taskId}}`, runId {{runId}}, name `{{name}}` not found',
+      { taskId, runId, name },
+    );
+  }
+
+  const artifact = artifactUtils.fromDb(artifact_row);
+  const run = task.runs[runId]; // checked to exist above
+
+  const publishCreatedMessage = async () => {
+    await this.publisher.artifactCreated({
+      status: task.status(),
+      artifact: artifactUtils.serialize(artifact),
+      workerGroup: run.workerGroup,
+      workerId: run.workerId,
+      runId,
+    }, task.routes);
+  };
+
+  // if this is of a type that is already finished, that's not allowed
+  if (artifact.storageType !== 'object') {
+    return res.resportError('InputError', 'Only object artifacts can be finished', {});
+  }
+
+  // If already marked present, this is an idempotent call, in which case
+  // we send the Pulse message again (at-least-once semantics)
+  if (artifact.present) {
+    await publishCreatedMessage();
+    return res.status(204).send();
+  }
+
+  if (uploadId !== artifact.details?.uploadId) {
+    return res.reportError(
+      'InputError',
+      'given uploadId does not match that for the artifact',
+      {});
+  }
+
+  // verify object is complete with object service, before allowing this
+  // operation to complete
+  try {
+    await this.objectService.object(artifactToObjectName(taskId, runId, name));
+  } catch (err) {
+    if (err.statusCode !== 404) {
+      throw err;
+    }
+    return res.reportError('InputError', 'Object storing artifact data does not exist', {});
+  }
+  await this.db.fns.queue_artifact_present(taskId, runId, name);
+  await publishCreatedMessage();
+
+  // clear out the details, since the uploadId is no longer relevant
+  await this.db.fns.update_queue_artifact_2({
```

Comment:
`update_queue_artifact_2` doesn't feel like an ideal function name - is that new in this PR, or something that already existed?

Reply:
It's been around for a while.  We started out using `_2`, .. when we modified DB function signatures, but have lately been using slightly more descriptive suffixes like `_tqid` or `_with_foo`.